### PR TITLE
Kinematics interface with KDL implementation

### PR DESCRIPTION
**Pull request by pac48**
Add kinematics interface with a KDL implementation

The proposed interface defines 5 public methods: `initialize`, `update_robot_state`, `convert_cartesian_deltas_to_joint_deltas`, `calculate_segment_transform`, and `convert_joint_deltas_to_cartesian_deltas`. These methods are sufficient to do basic end effector Cartesian control, but more methods would be needed if we wanted to also do forward and inverse dynamics.

The KDL implementation assumes a URDF has been loaded to the robot_description parameter.

One potential issue is that Eigen is allocating memory for its operations.

I would appreciate feedback anyone has (design pattern and naming conventions)

### CODE REVIEW

**Comment by dignakov** [comment]
I didn't say anything at the meeting, but I think this package is a great addition.

I do have one question/comment. There is already a place where the joint states are stored. What would be the reason for keeping a second set of state inside the kinematic interface? Could it not be a bit more functional where the joint positions, velocities, etc are passed into the functions?

**Comment by pac48** [comment]
> I didn't say anything at the meeting, but I think this package is a great addition.
> 
> I do have one question/comment. There is already a place where the joint states are stored. What would be the reason for keeping a second set of state inside the kinematic interface? Could it not be a bit more functional where the joint positions, velocities, etc are passed into the functions?

That is a good point. If the user is require to pass the robot state in all of the other functions, (`convert_cartesian_deltas_to_joint_deltas`, `calculate_segment_transform`, and `convert_joint_deltas_to_cartesian_deltas`), then `update_robot_state` can be removed completely and the interface is almost stateless. The interface can still be in two state of course: initialized/uninitialized. 

I guess the only real benefit to keeping the robot state internally is that the datatypes are different for each plugin. The values come in as vectors of doubles, but will need be copied to whatever type the kinematics interface uses. I am not sure that really matters that much.

**Comment by pac48** [comment]
I updated the interface to be stateless. Now each method requires you pass the joint positions in.

File: kdl_plugin/package.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>kdl_plugin</name>
```
Comment (bmagyar) [inline kdl_plugin/package.xml:4]: I'd suggest renaming this to `kinematics_interface_kdl` for more clarity. The name `kdl_plugin` is waaaay too general

File: kdl_plugin/package.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  <version>0.0.0</version>
+  <description>KDL based implementation of ROS2 control kinematics interface</description>
```
Comment (bmagyar) [inline kdl_plugin/package.xml:6]:
```suggestion
  <description>KDL based implementation of ros2_control kinematics interface</description>
```

File: kdl_plugin/kdl_plugin.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<library path="kdl_plugin">
+  <class name="kdl_plugin/KDLKinematics"
+         type="kdl_plugin::KDLKinematics"
+         base_class_type="kinematics_interface::KinematicsBaseClass">
+    <description>
+      KDL plugin for ROS2 control kinematics interface
```
Comment (bmagyar) [inline kdl_plugin/kdl_plugin.xml:6]:
```suggestion
      KDL plugin for ros2_control kinematics interface
```

**Review by bmagyar — CHANGES_REQUESTED** [review]
more to come tomorrow morning

File: kinematics_interface/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# find dependencies
+find_package(ament_cmake REQUIRED)
+find_package(geometry_msgs REQUIRED)
+find_package(rclcpp_lifecycle REQUIRED)
+
+
+# We require a differential inverse kinematics plugin for converting Cartesian admittance to joint admittance
+# Base class for the IK plugin
+add_library(kinematics_interface SHARED
+        src/kinematics_interface_base.cpp
```
Comment (bmagyar) [inline kinematics_interface/CMakeLists.txt:21]: am I correct that this file was only created to hack that header file into a library?

cmake has *interface libraries* for that: https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries

File: kinematics_interface/src/kinematics_interface_base.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+#include "kinematics_interface/kinematics_interface_base.hpp"
+
+namespace kinematics_interface
+{
+    KinematicsBaseClass::KinematicsBaseClass() {
+
+    }
+
+    //KinematicsBaseClass::~KinematicsBaseClass(){
```
Comment (bmagyar) [inline kinematics_interface/src/kinematics_interface_base.cpp:26]: please remove

File: kinematics_interface/package.xml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>kinematics_interface</name>
+  <version>0.0.0</version>
+  <description>Kinematics interface for ROS 2 control</description>
+  <maintainer email="denis@stogl.de">Denis Štogl</maintainer>
```
Comment (bmagyar) [inline kinematics_interface/package.xml:7]: can I sign up for membership here? (I'm already reviewing :D ) 
what's the plan for doing releases?

File: kdl_plugin/package.xml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>kinematics_interface_kdl</name>
+  <version>0.0.0</version>
+  <description>KDL implementation of ros2_control kinematics interface</description>
+  <maintainer email="paul.gesel@picknik.ai">paul</maintainer>
+  <license>TODO: License declaration</license>
```
Comment (bmagyar) [inline kdl_plugin/package.xml:8]: files say apache 2

**Review by bmagyar — CHANGES_REQUESTED** [review]

File: kdl_plugin/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,66 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (bmagyar) [inline kdl_plugin/CMakeLists.txt:1]: the folder name here still needs to be changed from `kdl_plugin`

**Review by bmagyar — APPROVED** [review]

**Pull request merged**